src/monitor/mod.rs

```diff
@@ -0,0 +1,570 @@
+                .ok_or_else(|| TraceError::UnknownNode(to.clone()))?;
+            self.circuit.add_edge(from, to, kind);
+            Ok(())
+        } else {
+            match event {
+                CircuitEvent::PushRegion { name } => {
+                    self.push_region(name.clone());
+                    Ok(())
+                }
+                CircuitEvent::PopRegion => self.pop_region(),
+                _ => panic!("unknown event"),
+            }
+        }
+    }
+
+    fn scheduler_event(&mut self, event: &SchedulerEvent) -> Result<(), TraceError> {
+        // eprintln!("scheduler event: {}", event);
+        if !self.running() {
+            return Err(TraceError::InvalidEvent(Cow::from(
+                "scheduler event received after circuit execution has terminated",
+            )));
+        }
+        let current_node_id = self.current_node_id();
+        match event {
+            SchedulerEvent::ClockStart => {
+                if !self
+                    .circuit
+                    .node_ref(&current_node_id)
+                    .unwrap()
+                    .is_iterative()
+                {
+                    return Err(TraceError::InvalidEvent(Cow::from(
+                        "received 'ClockStart' event for a non-iterative circuit",
+                    )));
+                }
+
+                if !self.current_state().is_idle() {
+                    return Err(TraceError::InvalidEvent(Cow::from(format!(
+                        "received 'ClockStart' event in state {}",
+                        self.current_state().name()
+                    ))));
+                }
+
+                self.set_current_state(CircuitState::Running);
+                Ok(())
+            }
+            SchedulerEvent::ClockEnd => {
+                if !self.current_state().is_running() {
+                    return Err(TraceError::InvalidEvent(Cow::from(format!(
+                        "received 'ClockEnd' event in state {}",
+                        self.current_state().name()
+                    ))));
+                }
+
+                self.pop_state();
+                Ok(())
+            }
+            SchedulerEvent::StepStart => {
+                if self
+                    .circuit
+                    .node_ref(&current_node_id)
+                    .unwrap()
+                    .is_iterative()
+                    && !self.current_state().is_running()
+                {
+                    return Err(TraceError::InvalidEvent(Cow::from(format!(
+                        "received 'StepStart' event in state {} of an iterative circuit",
+                        self.current_state().name()
+                    ))));
+                }
+                if !self
+                    .circuit
+                    .node_ref(&current_node_id)
+                    .unwrap()
+                    .is_iterative()
+                    && !self.current_state().is_idle()
+                {
```

Comment:
Pull this out of the conditions
```suggestion
                let current_node = self
                    .circuit
                    .node_ref(&current_node_id)
                    .unwrap();

                if current_node.is_iterative() && !self.current_state().is_running() {
                    return Err(TraceError::InvalidEvent(Cow::from(format!(
                        "received 'StepStart' event in state {} of an iterative circuit",
                        self.current_state().name()
                    ))));

                } else if !(current_node.is_iterative() && self.current_state().is_idle()) {
```